### ISSUE TITLE
Few improvements to docs for the ENI mode

### ISSUE THREAD

**Issue by errordeveloper** [cilium/cilium#10426]
## Proposal / RFE

**Is your feature request related to a problem?**

When using ENI mode, it's possible to run out of IPs. The [documentation](http://docs.cilium.io/en/latest/gettingstarted/aws-eni/) doesn't say anything about how one can go about troubleshooting it. It also doesn't specify what EC2 limits apply, especially with regards to max pods per-node for different instance types.

**Describe the solution you'd like**

- Document the symptoms:
   - `kubectl get pod` shows status `ContainerCreating`
   - `kubect describe pod` shows `unable to allocate IP via local cilium agent: [POST /ipam][502] postIpamFailure  No more IPs available` error message
- Document how to monitor and troubleshoot IP space usage
- Document what EC2 limits apply, reference https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt etc

**Comment by ungureanuvladvictor** [comment]
👍 to putting this in docs. We've had a lot of encounters with this internally many times -- usually due to ips leaking so some docs + example usage of metrics would go a long way for the average user.

**Comment by mvisonneau** [comment]
One thing which I believe might be common is that by default, the IPAM do not allocate IP addresses on the first interface. This is not the default behaviour with the aws-vpc-cni-k8s driver. The https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt does not reflect this and consequently, the kubelet will try to allocate more pods that Cilium can sustain.

2 solutions that I can think of:
- Configure the IPAM to use all interfaces by setting `spec.eni.first-interface-index` to **0**
- Update and use a `eni-max-pods.txt` which takes in account the lower amount of addresses available in order to let the kubelet knows

**Comment by mvisonneau** [comment]
a 3rd solution could be having a mechanism to fetch the amount of ip addresses available from the `CiliumNode` CRD and reconfigure the kubelet accordingly 🤷‍♂

**Comment by vietwow** [comment]
I got the same bug on EKS v1.15 with cillium 1.7.2 (installed by helm chart) :(

**Comment by mvisonneau** [comment]
I personally choose to use all interfaces for now, this can be done by using the following:

```
~$ kubectl apply -f - <<EOF
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: cilium-cni-configuration
  namespace: kube-system
data:
  cni-config: |
    {
      "cniVersion": "0.3.1",
      "name": "cilium",
      "type": "cilium-cni",
      "enable-debug": false,
      "eni": {
        "first-interface-index": 0
      }
    }
EOF
```

and overriding the helm chart values with: 

`--set global.cni.configMap=cilium-cni-configuration`

(thanks to @tgraf for the hint! 🙂)

**Comment by vietwow** [comment]
@mvisonneau it work like a charm, thank you so much

**Comment by olemarkus** [comment]
Think it is worth getting this documented. Ran into this myself.

**Comment by zhujik** [comment]
Yep, definitely, the DOC says to use the parameters but I had a lot of trouble finding out how, because there is no example for custom CNI configuration setup.

**Comment by krishadvani** [comment]
Unfortunetly this problem persists even after applying the cilium CNI config map, we tried running it on a t4g.medium instance, which supports 17 pods, cilium would end up scheduled 19 and stay at container creation state.

**Comment by gandro** [comment]
We've already changed the `"first-interface-index": 0` default in Cilium 1.8, so the CNI config is not doing anything anymore.

But there are two IPs reserved by Cilium itself, namely the `health` and `router` IP. That explains the difference of two pods not being able to be scheduled. Not sure what can be done about that other than documenting it better, since Cilium cannot change the max-pods limit by itself.

**Comment by krishadvani** [comment]
I understand. Any workarounds that are possible? Because this can cause the service to be unavailable if the pod we are trying to run is running two copies, and they both get scheduled in a node which will hold them to the container creation stage forever unless force deleted.

**Comment by krishadvani** [comment]
#21196

**Comment by gandro** [comment]
> I understand. Any workarounds that are possible? Because this can cause the service to be unavailable if the pod we are trying to run is running two copies, and they both get scheduled in a node which will hold them to the container creation stage forever unless force deleted.

I would recommend changing the `max-pods` default of your nodegroup. How to change this is documented here: https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html

The document explains how to increase the limit, but the steps for decreasing it are the same.

**Comment by krishadvani** [comment]
@gandro Appreciate your time on this. 

This does work, I wish there was a way to inform kubelet dynamically, but I don't think there is a way at this moment. I understand that there are 3 such IPs, health, router, and CiliumInternalIP. So doing a -3 on total allowed IPs is what helped us.

**Comment by sepich** [comment]
This link might be useful for docs, in terms of `max-pod-per-node` calculation, and increased instance limits in `prefix` mode:
https://aws.amazon.com/blogs/containers/amazon-vpc-cni-increases-pods-per-node-limits/

**Comment by krishadvani** [comment]
@sepich Does this work without CNI chaining in Cilium? I guess I will try it our, however would be great to know.

**Comment by sepich** [comment]
Yes, prefix assignment works in latest cilium right out of the box, without CNI chaining.

**Comment by krishadvani** [comment]
I read through the documentation, and these would only work with AWS Nitro instances, regular EC2 instances are out of the question. Further, if you don't use NodeGroups in your cluster, i.e if you use Karpenter autoscaler which works directly with EC2's then there is no way to specify max pods per node group either since node groups are not fixed unless you explicitly specify them while provisioning.

**Comment by christarazi** [comment]
Related: https://github.com/cilium/cilium/issues/22054

**Comment by jaffcheng** [comment]
This is more like a missing feature of kube-scheduler and not very much you can do from cilium. If adjusting `max-pods` is not sufficient, writing a custom [_Device Plugin_](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) could be an alternative solution.

![scheduling2](https://github.com/cilium/cilium/assets/9313189/d43336a0-a7d4-4f48-8c90-bd1e83931054)

pros:
- what you see from cilium-agent is what you get, regardless of the number of hostNetwork pods
- in case IP allocation stops working or not finished for new nodes, pre-allocated IPs would serve as a buffer before it becomes user-facing
- also works on other clouds as long as you use cilium-agent

cons:
- development & operation overhead

**Comment by Cbeck527** [comment]
I don't mean to resurrect an older thread, but we ran into this at our org and have worked around it with a little bit of guess and check:

Enabling prefix delegation[0] in the helm chart values files did not work for us, so we dug in a little bit. On the [cilium-operator docs page](https://docs.cilium.io/en/stable/cmdref/cilium-operator-aws/) there is a flag that you can pass to the operator `--aws-use-primary-address`. Passed this into the helm chart with `operator.extraArgs[]` and we were able to avoid the IP exhaustion issue since the primary ENI was now able to get more IPs.

[0] we ended up moving away from the single IP allocation entirely and got prefix delegation working. For whatever reason, the helm chart sets the value in the configmap but the operator doesn't pick it up. Using setting `--aws-enable-prefix-delegation` in the extra args worked like a charm. 🤷

**Comment by voltagebots** [comment]
> [0] we ended up moving away from the single IP allocation entirely and got prefix delegation working.
@Cbeck527 how did you move to a single IP allocation? looking at the [aws](https://aws.amazon.com/blogs/containers/amazon-vpc-cni-increases-pods-per-node-limits/) doc, it seems prefix delegation only works with AWS Nitro instance types.

**Comment by Cbeck527** [comment]
@voltagebots we were using C6 instances which are built on the nitro platform. It's somewhat hidden, but pretty much every instance type above generation 5 is built on the nitro platform.

https://docs.aws.amazon.com/ec2/latest/instancetypes/ec2-nitro-instances.html#nitro-instance-types
